### PR TITLE
Kafka transactions

### LINKED ISSUE #104: Support for Transactions
Please see [Transactions in Apache Kafka](https://www.confluent.io/blog/transactions-apache-kafka/) for the background on this.

Essentially, it would be a nice feature to support a transaction within the `KafkaConsumerProcessor`, especially with `@SendTo` with `transactional.id` producer configuration.

It seems that the `@SendTo` functionality closely models and resembles the `read-process-write` classification in the article.

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes #104

### CODE REVIEW

File: kafka/src/main/java/io/micronaut/configuration/kafka/KafkaProducerFactory.java [hunk 1]
```diff
@@ -163,10 +164,21 @@ public KafkaProducerFactory(BeanContext beanContext, SerdeRegistry serdeRegistry
...
+                properties.putIfAbsent(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+            }
+
             if (hasId) {
                 properties.putIfAbsent(ProducerConfig.CLIENT_ID_CONFIG, id);
             }
-            return beanContext.createBean(Producer.class, newConfig);
+
+            Producer producer = beanContext.createBean(Producer.class, newConfig);
+            if (transactional) {
```
Comment (graemerocher) [inline kafka/src/main/java/io/micronaut/configuration/kafka/KafkaProducerFactory.java:177]: These clients are cached, is it ok / thread safe to call `initTransaction` on a cached client?

File: kafka/src/main/java/io/micronaut/configuration/kafka/Message.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+import java.util.Map;
+
+/**
+ * Message payload representation.
+ *
+ * @author Denis Stepanov
+ * @since 4.1.0
+ */
+public final class Message {
```
Comment (graemerocher) [inline kafka/src/main/java/io/micronaut/configuration/kafka/Message.java:29]: Maybe a bit generic, call it `KafkaMessage`

File: kafka/src/main/java/io/micronaut/configuration/kafka/TransactionalProducerRegistry.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @NonNull
+    <K, V> Producer<K, V> getTransactionalProducer(@Nullable String clientId, String transactionalId, Argument<K> keyType, Argument<V> valueType);
+
+    /**
+     * Closed the producer.
+     * Should be used for cases when {@link org.apache.kafka.common.errors.ProducerFencedException} is thrown.
+     *
+     * @param producer The producer
+     */
+    void close(Producer<?, ?> producer);
```
Comment (graemerocher) [inline kafka/src/main/java/io/micronaut/configuration/kafka/TransactionalProducerRegistry.java:54]: Add `@NonNull`

**Review by graemerocher — CHANGES_REQUESTED** [review]
Awesome work! Just a few of changes

File: kafka/src/main/java/io/micronaut/configuration/kafka/KafkaProducerFactory.java (hunk 1 above)
Comment (dstepanov) [reply kafka/src/main/java/io/micronaut/configuration/kafka/KafkaProducerFactory.java:177]: My understanding is yes as it needs to be called once

**Review by graemerocher — APPROVED** [review]

**Pull request merged**